### PR TITLE
Use compressed fabric ID to advertise and resolve nodes on a fabric

### PR DESCRIPTION
**Pull request by pan-apple**
#### Problem
The mDNS advertisements and resolution is not using compressed fabric ID. That can cause collisions when multiple fabrics are commissioned on a device.

#### Change overview
Update `NodeId` to use compressed fabric ID.
Uncompressed IDs have limited usage (in CHIPCert, FabricTable and operational-credentials-server). Added a new class that is used only at these code blocks.

#### Testing
CI runs existing PeerId and Service Naming tests, that will now use compressed IDs.
Also tested commissioning flows using Python controller and chip-tools apps.

Python controller app's resolve command is impacted by this change. The command takes a fabric ID as the parameter. It should now be provided the compressed fabric ID value instead.
The compressed fabric ID can be retrieved by `get-fabricid` command.

### CODE REVIEW

File: examples/minimal-mdns/advertiser.cpp [hunk 1]
```diff
@@ -289,7 +289,7 @@ int main(int argc, char ** args)
                 .EnableIpV4(gOptions.enableIpV4)
                 .SetPort(CHIP_PORT)
                 .SetMac(chip::ByteSpan(gOptions.mac, 6))
-                .SetPeerId(PeerId().SetFabricId(gOptions.fabricId).SetNodeId(gOptions.nodeId)));
+                .SetPeerId(PeerId().SetCompressedFabricId(gOptions.fabricId).SetNodeId(gOptions.nodeId)));
```
Comment (tcarmelveilleux) [inline examples/minimal-mdns/advertiser.cpp:292]: In this example, setting direct compressed fabric ID seems odd and doesn't match what I would expect a tool user to do.

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp [hunk 2]
```diff
@@ -116,8 +116,8 @@ CHIP_ERROR writeFabricsIntoFabricsListAttribute()
     uint8_t fabricIndex = 0;
     for (auto & pairing : GetGlobalFabricTable())
     {
-        NodeId nodeId               = pairing.GetNodeId();
-        uint64_t fabricId           = pairing.GetFabricId();
+        NodeId nodeId               = pairing.GetPeerId().GetNodeId();
+        uint64_t fabricId           = pairing.GetUncompressedFabricId();
```
Comment (tcarmelveilleux) [inline src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:120]: Suggest using `GetRawFabricId()` as a better name.

Also, `writeFabric` is missing the root public key. It still seems like the code expects a single fabric space, and fabric ID collisions are still possible.

File: src/controller/java/CHIPDeviceController-JNI.cpp [hunk 3]
```diff
@@ -437,7 +437,7 @@ JNI_METHOD(void, updateDevice)(JNIEnv * env, jobject self, jlong handle, jlong f
     StackLockGuard lock(JniReferences::GetInstance().GetStackLock());
 
     AndroidDeviceControllerWrapper * wrapper = AndroidDeviceControllerWrapper::FromJNIHandle(handle);
-    CHIP_ERROR err = wrapper->Controller()->UpdateDevice(static_cast<chip::NodeId>(deviceId), static_cast<uint64_t>(fabricId));
+    CHIP_ERROR err                           = wrapper->Controller()->UpdateDevice(static_cast<chip::NodeId>(deviceId));
```
Comment (tcarmelveilleux) [inline src/controller/java/CHIPDeviceController-JNI.cpp:440]: These aligned equals are starting to really make me consider commenting on our style rules. Extending a line width that much doesn't help :( Not an issue for this PR, but we should consider not doing "="-aligment.

Comment (pan-apple) [reply src/controller/java/CHIPDeviceController-JNI.cpp:440]: For this PR, I can add a blank line, and that should remove this alignment change.

File: src/lib/core/PeerId.h [hunk 4]
```diff
@@ -39,15 +40,49 @@ class PeerId
...
+    PeerId & SetCompressedFabricId(CompressedFabricId id)
+    {
+        mCompressedFabricId = id;
+        return *this;
+    }
+
+    bool operator==(const PeerId & other) const
+    {
+        return (mNodeId == other.mNodeId) && (mCompressedFabricId == other.mCompressedFabricId);
+    }
```
Comment (tcarmelveilleux) [inline src/lib/core/PeerId.h:53]: This is very likely true, but not exactly true. We have to be careful, because if this is used in ACL processing one day, it is a security issue. There needs to be a match against root public key as well. I think PeerId could be refactored to always refer to a local fabric index, and have getters that index into fabrics info metadata to recover either the root public key, fabric ID or compressed fabric ID. This way, you could do full check against root as well.

File: src/lib/core/PeerId.h [hunk 5]
```diff
@@ -39,15 +40,49 @@ class PeerId
...
+    FabricId GetUncompressedFabricId() const { return mFabricId; }
+    UncompressedPeerId & SetUncompressedFabricId(FabricId id)
     {
         mFabricId = id;
         return *this;
     }
 
-    bool operator==(const PeerId & other) const { return (mNodeId == other.mNodeId) && (mFabricId == other.mFabricId); }
-    bool operator!=(const PeerId & other) const { return (mNodeId != other.mNodeId) || (mFabricId != other.mFabricId); }
+    bool operator==(const UncompressedPeerId & other) const { return (mNodeId == other.mNodeId) && (mFabricId == other.mFabricId); }
```
Comment (tcarmelveilleux) [inline src/lib/core/PeerId.h:84]: This also need to account for root, as previously mentioned.

File: examples/minimal-mdns/advertiser.cpp (hunk 1 above)
Comment (pan-apple) [reply examples/minimal-mdns/advertiser.cpp:292]: We'll need to provide the root pub key to the tool to compute the compressed fabric ID. That could be done, but need a clever way of grabbing the correct pub key, and be able to pass it to the command line tool.

I can file an issue and tag it with this change..

File: src/transport/FabricTable.cpp [hunk 6]
```diff
@@ -337,18 +350,21 @@ CHIP_ERROR FabricInfo::VerifyCredentials(const ByteSpan & opCertArray, Validatio
...
+    UncompressedPeerId uncompressedPeerId;
+    ReturnErrorOnFailure(ExtractPeerIdFromOpCert(certificates.GetCertSet()[nocCertIndex], &uncompressedPeerId));
 
     if (!icac.empty())
     {
         FabricId icacFabric;
         if (ExtractFabricIdFromCert(certificates.GetCertSet()[1], &icacFabric) == CHIP_NO_ERROR)
         {
-            ReturnErrorCodeIf(icacFabric != nocPeerId.GetFabricId(), CHIP_ERROR_INVALID_CASE_PARAMETER);
+            ReturnErrorCodeIf(icacFabric != uncompressedPeerId.GetUncompressedFabricId(), CHIP_ERROR_INVALID_CASE_PARAMETER);
```
Comment (tcarmelveilleux) [inline src/transport/FabricTable.cpp:361]: This changes the meaning of the check.

It's possible ICAC doesn't have fabric ID attribute. I think the prior behavior would have yielded 0, and same for the default constructed FabricId (unless there was a bug....). Overall, this check must only be done if the ICAC does indeed have a chip-fabric-id. RDN in its subject

File: src/transport/FabricTable.cpp [hunk 7]
```diff
@@ -170,6 +173,16 @@ CHIP_ERROR FabricInfo::FetchFromKVS(PersistentStorageDelegate * kvs)
     return err;
 }
 
+CHIP_ERROR FabricInfo::GetCompressedId(const UncompressedPeerId & uncompressedPeerId, PeerId & compressedPeerId) const
```
Comment (tcarmelveilleux) [inline src/transport/FabricTable.cpp:176]: Suggest making clear in argument name that comrpessedPeerId is an output

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 2 above)
Comment (pan-apple) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp:120]: `GetRawFabricId()` sounds good.

`writeFabric` needs to be updated. The `FabricDescriptor` structure is out of date compared to the latest spec.
It's also missing `FabricIndex`, and ordering of the fields is not correct. I am planning a separate PR for this part of the code.

File: src/lib/core/PeerId.h (hunk 4 above)
Comment (pan-apple) [reply src/lib/core/PeerId.h:53]: Is the concern that there could be different combinations of `<uncompressed Fabric ID, Root pubkey>` that can yield the same compressed fabric ID?

Comment (pan-apple) [reply src/lib/core/PeerId.h:53]: I tried making `PeerId` carry a `P256Pubkey` in it (for the root pub key). That's causing circular dependencies, unfortunately. So we'll need to move `PeerId.h` out of `src/lib/core` to some module that's already dependent on `src/Crypto`.

I'll submit an issue to track it.

Comment (pan-apple) [reply src/lib/core/PeerId.h:54]: https://github.com/project-chip/connectedhomeip/issues/9192

File: src/lib/core/PeerId.h (hunk 5 above)
Comment (pan-apple) [reply src/lib/core/PeerId.h]: https://github.com/project-chip/connectedhomeip/issues/9192

File: examples/minimal-mdns/advertiser.cpp (hunk 1 above)
Comment (pan-apple) [reply examples/minimal-mdns/advertiser.cpp:292]: https://github.com/project-chip/connectedhomeip/issues/9193

File: src/app/clusters/operational-credentials-server/operational-credentials-server.cpp (hunk 2 above)
Comment (pan-apple) [reply src/app/clusters/operational-credentials-server/operational-credentials-server.cpp]: https://github.com/project-chip/connectedhomeip/issues/9194

File: src/lib/core/PeerId.h (hunk 4 above)
Comment (tcarmelveilleux) [reply src/lib/core/PeerId.h:54]: If `src/Crypto` depends on `PeerId`, there is an abstraction problem and we can resolve it.

The PeerID doesn't need to carry the key itself. It can carry FabricId, and that's used to later lookup a key if/when needed

And re:

> Is the concern that there could be different combinations of <uncompressed Fabric ID, Root pubkey> that can yield the same compressed fabric ID?

The answer is yes. For discovery, it's not a problem, since the worst case is a denial of service. For access control  enforcement of subject or CASE, the tuple <Root public key, uncompressed fabric id, node id> is what must explicitly be matched (somehow, not necessarily through the bits, but at least in a way that it's not spoofable/collidable).

File: examples/minimal-mdns/advertiser.cpp (hunk 1 above)
Comment (tcarmelveilleux) [reply examples/minimal-mdns/advertiser.cpp:292]: Suggest that commissioning context be configurable in tools so that for dev testing, you can "name"/"alias" a fabric to its combination of <root cert, fabric id> and also configure the dev ICA (if any) for a given alias for exemplary credential issuance. This would go a very long way in help support multi-fabric testing.

File: src/transport/FabricTable.cpp [hunk 8]
```diff
@@ -173,13 +173,14 @@ CHIP_ERROR FabricInfo::FetchFromKVS(PersistentStorageDelegate * kvs)
     return err;
 }
 
-CHIP_ERROR FabricInfo::GetCompressedId(const UncompressedPeerId & uncompressedPeerId, PeerId & compressedPeerId) const
+CHIP_ERROR FabricInfo::GetCompressedId(const RawPeerId & rawPeerId, PeerId * compressedPeerId) const
```
Comment (tcarmelveilleux) [inline src/transport/FabricTable.cpp:176]: When I suggested making the argument clearer it's an output, I didn't think changing away from reference was necessary, but maybe renaming `outCompressedPeerId` or such. Making it a pointer also helps though, so this is satisfactory as-is I would say.

File: src/transport/FabricTable.cpp [hunk 9]
```diff
@@ -350,21 +351,22 @@ CHIP_ERROR FabricInfo::VerifyCredentials(const ByteSpan & opCertArray, Validatio
...
     if (!icac.empty())
     {
-        FabricId icacFabric;
-        if (ExtractFabricIdFromCert(certificates.GetCertSet()[1], &icacFabric) == CHIP_NO_ERROR)
+        FabricId icacFabric = kUndefinedRawFabricId;
+        if (ExtractFabricIdFromCert(certificates.GetCertSet()[1], &icacFabric) == CHIP_NO_ERROR &&
+            icacFabric != kUndefinedRawFabricId)
         {
-            ReturnErrorCodeIf(icacFabric != uncompressedPeerId.GetUncompressedFabricId(), CHIP_ERROR_INVALID_CASE_PARAMETER);
+            ReturnErrorCodeIf(icacFabric != rawPeerId.GetRawFabricId(), CHIP_ERROR_INVALID_CASE_PARAMETER);
```
Comment (tcarmelveilleux) [inline src/transport/FabricTable.cpp:363]: Would be great if we had an error that made this abundantly clear, like `CHIP_ERROR_FABRIC_MISMATCH_ON_ICA`, which may occur in a variety of places (including UpdateNOC command handling)

File: src/controller/python/chip-device-ctrl.py [hunk 10]
```diff
@@ -855,14 +855,15 @@ def do_getfabricid(self, line):
                 print("Unexpected argument: " + args[1])
                 return
 
-            fabricid = self.devCtrl.GetFabricId()
+            fabricid = self.devCtrl.GetCompressedFabricId()
```
Comment (tcarmelveilleux) [inline src/controller/python/chip-device-ctrl.py:858]: Suggest adding a getter for uncompressed fabric ID as well

File: src/controller/python/chip-device-ctrl.py [hunk 11]
```diff
@@ -855,14 +855,15 @@ def do_getfabricid(self, line):
...
         except exceptions.ChipStackException as ex:
             print("An exception occurred during reading FabricID:")
             print(str(ex))
             return
 
         print("Get fabric ID complete")
-        print("Fabric ID: " + hex(fabricid))
+        print("Compressed Fabric ID: " +
+              hex(fabricid) + " (" + str(fabricid) + ")")
```
Comment (tcarmelveilleux) [inline src/controller/python/chip-device-ctrl.py:866]: Suggest using `%08X` or similar, to ensure format is always 0-padded 64-bit hex value. Suggest also including the raw fabric ID, so that when root public key is more cleanly handled, all items can be provided: <root pk, fabric id> and compressed fabric ID

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by pan-apple** [comment]
rebased due to ZAP failures.

**Comment by pan-apple** [comment]
@saurabhst, @jelderton, do you have any review feedback?

**Review by woody-apple — APPROVED** [review]

**Review by tcarmelveilleux — APPROVED** [review]

**Comment by woody-apple** [comment]
@saurabhst @Damian-Nordic @msandstedt @mspang ?

**Review by andy31415 — APPROVED** [review]

File: examples/chip-tool/commands/common/Command.h [hunk 12]
```diff
@@ -109,7 +109,7 @@ class Command
     Command(const char * commandName) : mName(commandName) {}
     virtual ~Command() {}
 
-    void SetExecutionContext(ExecutionContext & execContext) { mExecContext = &execContext; }
+    void SetExecutionContext(ExecutionContext & execContext) { memcpy(&mExecContext, &execContext, sizeof(execContext)); }
```
Comment (msandstedt) [inline examples/chip-tool/commands/common/Command.h:112]: Why this change?

File: src/app/server/Mdns.cpp [hunk 13]
```diff
@@ -116,19 +116,18 @@ CHIP_ERROR AdvertiseOperational()
...
-            const auto advertiseParameters =
-                chip::Mdns::OperationalAdvertisingParameters()
-                    .SetPeerId(PeerId().SetFabricId(fabricInfo.GetFabricId()).SetNodeId(fabricInfo.GetNodeId()))
-                    .SetMac(FillMAC(mac))
-                    .SetPort(GetSecuredPort())
-                    .SetMRPRetryIntervals(CHIP_CONFIG_MRP_DEFAULT_ACTIVE_RETRY_INTERVAL,
-                                          CHIP_CONFIG_MRP_DEFAULT_ACTIVE_RETRY_INTERVAL)
-                    .EnableIpV4(true);
+            const auto advertiseParameters = chip::Mdns::OperationalAdvertisingParameters()
+                                                 .SetPeerId(fabricInfo.GetPeerId())
```
Comment (msandstedt) [inline src/app/server/Mdns.cpp:120]: Based on the code before this PR, this is probably functionally correct.  However, why are we advertising something called peer ID?  That makes no sense.

File: src/controller/CHIPDeviceController.h [hunk 14]
```diff
@@ -291,13 +290,14 @@ class DLL_EXPORT DeviceController : public Messaging::ExchangeDelegate,
...
-     * @return CHIP_ERROR CHIP_NO_ERROR on success, or corresponding error code.
+     * @brief Get the Compressed Fabric ID assigned to the device.
      */
-    CHIP_ERROR GetFabricId(uint64_t & fabricId);
+    uint64_t GetCompressedFabricId() const { return mLocalId.GetCompressedFabricId(); }
+
+    /**
+     * @brief Get the raw Fabric ID assigned to the device.
+     */
+    uint64_t GetRawFabricId() const { return mRawFabricId; }
```
Comment (msandstedt) [inline src/controller/CHIPDeviceController.h:300]: I think the 'raw fabric ID' terminology is well-meaning and an attempt for symmetry with 'compressed fabric ID'.

However, all of this is confusing enough and the spec has already defined terms for these things:

* fully-qualified fabric reference: the <root CA public key, fabric ID> tuple
* fabric ID: self-explanatory
* compressed fabric ID/identifier: the 64-bit hash of <root CA public key, fabric ID> we're now using in this PR

Given this, can we just stick to `FabricId` instead of `RawFabricId`?  This is the terminology the spec uses and that should be our guide.

File: src/lib/core/PeerId.h [hunk 15]
```diff
@@ -39,19 +41,53 @@ class PeerId
...
+        return (mNodeId != other.mNodeId) || (mCompressedFabricId != other.mCompressedFabricId);
+    }
+
+private:
+    NodeId mNodeId = kUndefinedNodeId;
+
+    CompressedFabricId mCompressedFabricId = kUndefinedFabricId;
+};
+
+class RawPeerId
```
Comment (msandstedt) [inline src/lib/core/PeerId.h:66]: Why do we need this?  This is not a raw node identifier; it's an incomplete node identifier.  If we do need it, can we have a TODO and a ticket?  This needs to go away.

Also, suggest a name like `IncompleteNodeIdentifier` or similar so nobody mistakes this for a guaranteed-unique key; it's not.

File: examples/chip-tool/commands/common/Command.h (hunk 12 above)
Comment (msandstedt) [reply examples/chip-tool/commands/common/Command.h:112]: Explained in slack.

We need a deep copy of mExecContext.  Still, suggest this:

```suggestion
    void SetExecutionContext(ExecutionContext & execContext) { mExecContext = execContext; }
```

File: src/app/server/Mdns.cpp (hunk 13 above)
Comment (msandstedt) [reply src/app/server/Mdns.cpp:120]: Discussed offline.  This object name (PeerId) might be confusing, but is preexisting.  Maybe in the future we can come up with a better name to describe what is in essence a fully-qualified node identity, which is obviously different than NodeId.

**Review by msandstedt — APPROVED** [review]
